File: 2023/09.lua
require "util"

function allZeros (history) 
    for _,e in pairs(history) do
        if e ~= 0 then return false end
    end
    return true
end
function sequencer (history)
    local last = {}
    local h = history
    while true do
       table.insert(last, h[#h]) 
       local nh = {}
       for i=2, #h do
           table.insert(nh, h[i] - h[i-1])
       end
       h = nh
       if allZeros (h) then break end
    end
    local seqSum = 0
    for _, e in pairs(last) do
        seqSum = seqSum + e
    end
    return seqSum
end 

function partOne (lines) 
    local sum = 0
    for k,history in pairs(lines) do
        ans = sequencer(history) 
        sum = sum + ans
    end
    return sum
end

function partTwo (lines) 
    return 'todo'
end

function main() 
    lines = {}
    for line in io.lines() do
        nums = {}
        -- k** regex! vooral de negatives niet vergeten 
        for s in line:gmatch("(-?%d+) ?") do
            table.insert(nums,tonumber(s))
        end
        table.insert(lines, nums)
    end
    print('partOne', partOne(lines))
    print('partTwo', partTwo(lines))

end
main()

File: 2023/09.js
const R = require('ramda')
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })
const tap = R.tap
const log = console.log // eslint-disable-line
const sum = (a,b) => a+b // eslint-disable-line
const mul = (a,b) => a*b // eslint-disable-line
const int = R.pipe(R.trim, parseInt) // eslint-disable-line

function main () {
  const lines = []

  rl.on('line', line => {
    line = line.split(' ').map(int)

    lines.push(line)
  })

  rl.on('close', () => {
    console.log('partOne', partOne(lines))
    console.log('partTwo', partTwo(lines))
  })
}

main()

function sequencer (history) {
  const last = []
  let h = [...history]
  while (true) {
    last.push(h[h.length - 1])

    const nh = []
    for (let i = 1; i < h.length; i += 1) {
      nh.push(h[i] - h[i - 1])
    }
    h = nh
    if (h.every(n => n === 0)) break
  }
  return last.reduce(sum)
}

function partOne (lines) {
  return lines
    // .map(tap(log))
    .map(sequencer).reduce(sum)
}

function partTwo (lines) {
  return 'todo'
}
